docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,157 @@
+# Overview
+Currently cascading deletion is implemented at client-side. For example, when deleting a replica set, kubectl uses a reaper to delete the created pods and then delete the replica set. We plan to move the cascading deletion to the server to simplify the client-side logic. In this proposal, we present two alternative designs. [Design #1](#design-1-tombstone--gc) uses the tombstone mechanism, and [Desig #2](#design-2-finalizer-framework--gc) uses the finalizer mechanism.
+
+# Goals
+* Supporting cascading deletion at the server-side. 
+* Centralizing the cascading deletion logic, rather than spreading in controllers.
+* Allowing optionally orphan the dependent objects.
+
+# Non goals
+* Releasing the name of an object immediately, so it can be reused ASAP.
+
+# Design 1. Tombstone + GC
+## API Changes
+```
+type DeleteOptions struct {
+	…
+	Orphaning bool
+}
+```
+**DeleteOptions.Orphaning**: allows a user to express whether the child objects should be orphaned.
+
+```
+type ObjectMeta struct {
+	...	
+	ParentReferences []ObjectReference
+}
+```
+**ObjectMeta.ParentReferences**: links the resource to the parent resources. For example, a replica set `R` created by a deployment `D` should have an entry in ObjectMeta.ParentReferences pointing to `D`. The link should be set when the child object is created. It can be updated after the creation.
+
+```
+type Tombstone struct {
+	UID types.UID
+}
+```
+**Tombstone**: a tombstone is created when an object is deleted and the user requires the children to be orphaned.
+**Tombstone.UID**: the UID of the original object.  
+
+## New components
+The only new component is the Garbage Collector, which consists of a scanner, a garbage processor, and a propagator.
+* Scanner:
+  * Uses the discovery API to detect all the resources supported by the system.
+  * Periodically scans all resources in the system and adds each object to the *Dirty Queue*.
+
+* Garbage Processor:
+  * Consists of the *Dirty Queue* and workers.
+  * Each worker:
+    * Dequeues an item from *Dirty Queue*
+    * Checks if the item’s parents still exist. 
+    * Checks for a tombstone of parents that don’t exist
+    * If there is no parent nor tombstone, requests the API server to delete the item.
+    * If tombstone does exist, removes the parent from the object’s ParentReferences.
+
+* Propagator:
+  * The Propagator is for optimization, not for correctness.
+  * Maintains a DAG of parent-child relations. This DAG stores only name/uid/orphan triplets, not the entire body of every item.
+  * Consists of an *Event Queue* and a single worker.
+  * Watches for create/update/delete events for all resources, enqueues the events to the *Event Queue*.
```

Comment:
OpenShift serves it's non-Kubernetes resources under the /oapi path so there are some problems using the discovery client right now that we need to work through.  I think we would need to run two garbage collectors.  The kubernetes garbage collector would probably run with a different identity than the garbage collector that deletes OpenShift resources.  For example, the current namespace controller is run twice in OpenShift, one that has authority to delete kube resources, the other that has authority to delete OpenShift resources.